design_drafts/controller_execution_management.md

```diff
@@ -0,0 +1,125 @@
+
+### Parallel Controller Container
+
+Just like in the sequential execution, we propose having a container structure which holds a set of loaded controllers which are being executed in parallel and primarily independent from each other.
+When all controllers were executed, each individual calculated values are being joined.
+As already mentioned previously, this parallel container can be initialized with a custom designed operation which dictates how all commonly calculated values (e.g. `torques`) are being combined.
+
+The parallel container also has only one input and one output state.
+However, the input state has to be multiplexed (essentially copied) for each controller and can't be shared in order to guarantee and isolated execution where no other controller within the same container can interfere.
+
+The image below shows how this execution model could look like:
+![parallel_execution](images/parallel_execution.png "Parallel Execution Container")
+
+## Nested Controllers
+
+In both cases, the sequential as well as the parallel container can be controllers in itself.
+That gives the flexibility to easily nest the container within another container transparently.
+Again, a very simple scenario could be a parallel `PID` controller setup which is then being configured within a sequential container to enforce joint limits.
+This further allows to consider maybe very complex configured containers as black boxes, yet have the freedom to extend these fairly easy.
+
+The image below shows how this execution model could look like:
+![nested_execution](images/nested_execution.png "Nested Execution Container")
+
+## Variable Controller Frequencies
+
+One big constraint of the round-robin approach is the constant update frequency for all loaded controllers.
+There are plenty of use-cases however where it makes sense to have variable update frequencies, such as running a robot controller with 1 kHz whereas a camera controller only receives new images with a 20 Hz frequency.
+Every controller should therefore be attached with its own update frequency.
+
+Looking at the `rclcpp executor` model, one approach could be to start a timer for each controller where its callback will call the respective `update()` function.
+In order to establish a deterministic behavior of the timers, the controller manager has to make sure that really only timers are started and no other blocking calls such as subscriptions or services are part of it.
+There is quite some discussion about this at the moment of writing here: https://answers.ros.org/question/327477/ros2-uses-6-times-more-cpu-than-fastrtps/
+
+The potential options are having a separate executor designed for ros control with adheres to the interface of the `rclcpp::executor` or to neglect the concept and implement a custom execution model for ros_control.
+
+## Implementation - Proof of Concept
```

Comment:
Let's add a new section or something to this elaborating on ideas for existing/new tooling that could be used to assess the state of a system in configuration time, runtime, etc.

For configuration I can see a tool reading configuration files from a launch file perhaps that was meant to bring everything up (this way not all controller config files need to live in the same place) and construct a schematic of the system with a simulink-style gui?

Reply:
yes! I very much like the idea to have some sort of human readable configuration file (.yaml) or such which designs the controller setup. That would also allow to even have bindings to graphical user interfaces which generate these config files.